Have the font used for context sentences change size depending on accessibility options.

kyuuzou opened the pull request:
Content sentences are too small for my aging eyesight. This pull request scales the font size depending on the accessibility settings under:
Display & Text Size -> Larger Text 

![Screenshot 2023-05-19 at 15 44 32](https://github.com/davidsansome/tsurukame/assets/2806093/aac87fc2-212e-48a8-8049-cd98a81c91b7)

Fixes #624

tallerasaf commented:
@kyuuzou @davidsansome 

Is this PR solving, those two issues?
1. [](https://github.com/davidsansome/tsurukame/issues/334)
2. [](https://github.com/davidsansome/tsurukame/issues/601)

kyuuzou commented:
@tallerasaf More or less. It does make the size of the text on the context sentences adapt to the size selected on the accessibility options, which addresses some of the concerns on those issues, but it doesn't solve the problem on those screenshots that show that if you set the font size really high, then the layout becomes broken.

tallerasaf commented:
> It doesn't solve the problem on those screenshots that show that if you set the font size really high, then the layout becomes broken.

@kyuuzou @davidsansome Is there an easy way to fix the font for the entire app that it will look good on any accessibility font size? While keeping the layout stable on every font.

tallerasaf commented:
To scale the layout and font size of your iPhone app based on the Accessibility Settings, you can use Dynamic Type. Dynamic Type automatically adjusts the font size according to the user's preference. Here’s how to do it:

1. **Use System Fonts**: In your Interface Builder, set the fonts to use Text Styles (e.g. Headline, Body). For programmatically created labels, set their font like this:
    ```swift
    label.font = UIFont.preferredFont(forTextStyle: .body)
    ```

2. **Adjust Automatically**: Allow your labels and text elements to automatically adjust. In Interface Builder, check the “Automatically Adjusts Font” box for each label. Programmatically, you can do:
    ```swift
    label.adjustsFontForContentSizeCategory = true
    ```

3. **Listen for Changes**: If you need to do additional layout changes when text size changes, listen for `UIContentSizeCategory.didChangeNotification`. Example:
    ```swift
    NotificationCenter.default.addObserver(self, selector: #selector(handleTextSizeChange), name: UIContentSizeCategory.didChangeNotification, object: nil)
    ```
    And then in your function:
    ```swift
    @objc func handleTextSizeChange(notification: NSNotification) {
        // Perform your layout changes here
    }
    ```

4. **Use Adaptive Layout**: Ensure your Auto Layout constraints are set up in a way that they can accommodate the text size changes.

5. **Support Per-App Settings**: iOS 15 introduced per-app settings for accessibility. To make sure your app supports this feature, link against iOS 15 or later SDK and include the key `UIApplicationSupportsPreferredContentSizeCategoryPerApplication` in your app's Info.plist file:
    ```xml
    <key>UIApplicationSupportsPreferredContentSizeCategoryPerApplication</key>
    <true/>
    ```

By doing the above, your app should scale its font sizes and layout according to the user's Accessibility Settings for "Display & Text Size".


In addition to what I previously mentioned, there are a few more considerations to ensure that your app scales appropriately with the iPhone Accessibility Settings:

1. **Custom Fonts**: If you're using custom fonts, make sure to scale them properly using `UIFontMetrics`. For example:
    ```swift
    let font = UIFont(name: "CustomFontName", size: 17)!
    let scaledFont = UIFontMetrics.default.scaledFont(for: font)
    label.font = scaledFont
    label.adjustsFontForContentSizeCategory = true
    ```

2. **Images**: If you have images that need to scale with text, consider using `UIImageSymbolConfiguration` for SF Symbols, or make custom resizable images.

3. **Text Containers**: For elements like `UITextView` ensure that they are configured to update automatically as well. Example:
    ```swift
    textView.font = UIFont.preferredFont(forTextStyle: .body)
    textView.adjustsFontForContentSizeCategory = true
    ```

4. **Testing and Validation**: It's important to thoroughly test the app using different text sizes. Use different devices or simulators and change the text size in the settings to see how your app responds.

5. **Spacing and Constraints**: Pay attention to spacing between elements. Sometimes increasing font size might cause overlapping or truncation. Make sure your Auto Layout constraints are flexible.

6. **Customization for Larger Sizes**: Sometimes, simply scaling might not be enough, and you might want to provide a different layout or design for larger text sizes. Detect the current content size category and make changes accordingly:
    ```swift
    if traitCollection.preferredContentSizeCategory.isAccessibilityCategory {
        // Custom layout for larger sizes
    }
    ```

By taking into account these considerations in addition to the steps previously mentioned, you can create an app that is responsive and accessible to users with varying text size preferences.

davidsansome commented:
Did you just ask ChatGPT how to do it and paste its answer in here?

tallerasaf commented:
> Did you just ask ChatGPT how to do it and paste its answer in here?

Yes, sorry it was just for reference.
